chemprop/v2/data/utils.py

```diff
@@ -1,66 +1,163 @@
+            result = mol_split_fun(np.array(mols_without_atommaps), sampler="scaffold", **astartes_kwargs)
+            train, val, test = _unpack_astartes_result(datapoints, result, include_val)
+
+        # Use to constrain data with the same smiles go in the same split.
+        case SplitType.RANDOM_WITH_REPEATED_SMILES:
+            # get two arrays: one of all the smiles strings, one of just the unique
+            all_smiles = np.array([Chem.MolToSmiles(d.mol) for d in datapoints])
+            unique_smiles = np.unique(all_smiles)
+
+            # save a mapping of smiles -> all the indices that it appeared at
+            smiles_indices = {}
+            for smiles in unique_smiles:
+                smiles_indices[smiles] = np.where(all_smiles == smiles)[0]
+
+            # randomly split the unique smiles
+            result = split_fun(np.arange(len(unique_smiles)), sampler="random", **astartes_kwargs)
+            train_idxs, val_idxs, test_idxs = _unpack_astartes_result(None, result, include_val)
 
+            # convert these to the 'actual' indices from the original list using the dict we made
+            train = [datapoints[ii] for ii in itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in train_idxs)]
+            val = [datapoints[ii] for ii in itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in val_idxs)]
+            test = [datapoints[ii] for ii in itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in test_idxs)]
+
+        case SplitType.RANDOM:
+            result = split_fun(np.arange(len(datapoints)), sampler="random", **astartes_kwargs)
+            train, val, test = _unpack_astartes_result(datapoints, result, include_val)
+
+        case SplitType.KENNARD_STONE:
+            result = mol_split_fun(
+                np.array([d.mol for d in datapoints]),
+                sampler="kennard_stone",
+                hopts=dict(metric="jaccard"),
+                fingerprint="morgan_fingerprint",
+                fprints_hopts=dict(n_bits=2048),
+                **astartes_kwargs,
+            )
+            train, val, test = _unpack_astartes_result(datapoints, result, include_val)
+
+        case SplitType.KMEANS:
+            result = mol_split_fun(
+                np.array([d.mol for d in datapoints]),
+                sampler="kmeans",
+                hopts=dict(metric="jaccard"),
+                fingerprint="morgan_fingerprint",
+                fprints_hopts=dict(n_bits=2048),
+                **astartes_kwargs,
+            )
+            train, val, test = _unpack_astartes_result(datapoints, result, include_val)
+
+        case _:
+            raise ValueError(f'split type "{split}" not supported.')
+
+    return train, val, test
+
+
+def _unpack_astartes_result(
+    data: Sequence[MoleculeDatapoint], result: tuple, include_val: bool
+) -> Tuple[list[MoleculeDatapoint], list[MoleculeDatapoint], list[MoleculeDatapoint]]:
```

Comment:
This still has a capital `Tuple`. It can be lowercased and then import Tuple removed. I'm also happy to just merge the PR and not worry about this.